platform-sdk/docs/proposals/state-lifecycle/proposal.md

```diff
@@ -0,0 +1,231 @@
+ * becomes explicit and has a reservation count of 1. After this point in time, the reservation remains explicit
+ * until the object is eventually destroyed. It is impossible for an object with an explicit reservation to
+ * return to having an implicit reservation. If at any time after the object obtains an explicit reservation it
+ * has {@link #release()} called enough times to reduce its reference count to 0, then that object is destroyed
+ * and has its reference count set to -1.
+ */
+public interface Reservable extends Releasable {
+
+    /**
+     * The reference count of an object with an implicit reference.
+     */
+    int IMPLICIT_REFERENCE_COUNT = 0;
+
+    /**
+     * The reference count of an object with an explicit reference.
+     */
+    int DESTROYED_REFERENCE_COUNT = -1;
+
+    /**
+     * Acquire a reservation on this object. Increments the reference count by 1.
+     *
+     * @throws com.swirlds.common.exceptions.ReferenceCountException
+     * 		if this object has been fully released and destroyed
+     */
+    void reserve();
+
+    /**
+     * Attempts to acquire a reservation on this object. If the object is destroyed, the reservation attempt will fail.
+     *
+     * @return true if a reservation was acquired.
+     */
+    boolean tryReserve();
+
+    /**
+     * <p>
+     * Release a reservation on an object. Decrements the reference count by 1. If this method releases
+     * the last reservation, then this object should be destroyed.
+     * </p>
+     *
+     * <p>
+     * Should be called exactly once for each time {@link #reserve()} is called. The exception to this rule is
+     * if this object only has an implicit reference (i.e. {@link #getReservationCount()} returns
+     * {@link #IMPLICIT_REFERENCE_COUNT}). An object has an implicit reference immediately after it is constructed
+     * but before {@link #reserve()} has been called the first time. If called with an implicit reference, this
+     * object will be destroyed.
+     * </p>
+     *
+     * @return true if this call to release() caused the object to become destroyed
+     * @throws com.swirlds.common.exceptions.ReferenceCountException
+     * 		this object has already been fully released and destroyed
+     */
+    @Override
+    boolean release();
+
+    /**
+     * Get the total number of times {@link #reserve()} has been called minus the number of times {@link #release()}
+     * has been called. Will return {@link #IMPLICIT_REFERENCE_COUNT} if {@link #reserve()} has never been called,
+     * or {@link #DESTROYED_REFERENCE_COUNT} if this object has been fully released.
+     *
+     * @return
+     */
+    int getReservationCount();
+}
+
+```
+
+We're going to rely on this mechanism to manage the number of states that are kept in memory. 
+
+**Important**: It's up to the client code to make sure that the state is released once it's no longer needed.
```

Comment:
Will the State Module ensure that a state is at least hashed before it is garbage collected even if it's reservations are at 0? Or is it entirely on the client to maintain reservations in all cases?

Reply:
I see two possible approaches to that: 
- It may be entirely up to client to maintain reservations and make sure that they are acquired and released properly. In this case, `StateLifecycleManager` will not track if the state is hashed or not. 
- Reservation maintenance may be responsibility of `StateLifecycleManager`, and direct reservation of `State` instance should not be allowed. In this case, `StateLifecycleManager` has more control over `State` and can make sure that it's hashed before it's released but in this case we would have to maintain a list of `State` instances in memory. 

I'm leaning towards option 1.